### PR TITLE
Spy xc 2

### PR DESCRIPTION
**Pull request by mbarekh**
support route change and update webpack

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -52,6 +52,7 @@
     "react": "^15.4.1",
     "react-dom": "^15.4.1",
     "react-redux": "^5.0.2",
+    "react-router-dom": "^4.0.0-beta.8",
```
Comment (nicolaserny) [inline package.json]: you can add types too
npm install --save @types/react-router-dom

File: src/components/Components.tsx [hunk 2]
```diff
@@ -14,33 +14,38 @@ import { Dispatch } from "redux";
 import { XCSpyState } from "reducers/SpyReducer";
 import { ComponentProperties } from "reducers/components";
 import * as Box from "grommet/components/Box";
+import { withRouter } from "react-router-dom";
+import {API, CURRENT_COMPONENT, SERVER_URL } from "utils/urlParams";
 
 interface ComponentsGlobalProps extends ComponentsProps, ComponentsCallbackProps {
-    compositionModel: any;
+    history: any;
```
Comment (nicolaserny) [inline src/components/Components.tsx]: specify the type

File: src/components/SideBar.tsx [hunk 3]
```diff
@@ -75,22 +80,19 @@ class SideBar extends React.Component<SideBarGlobalProps, XCSpyState> {
...
-                <Anchor
-                    primary={true}
+                <Link
                     key={component}
                     value={component}
-                    className={(props.currentComponent === component) ? "active" : ""}
-                    onClick={(e) => {
-                        props.setCurrentComponent(e.target.getAttribute("value"));
-                    }}>
+                    to={{ pathname: "/app/", search: `${SERVER_URL}=${props.serverUrl}&${API}=${props.api}&${CURRENT_COMPONENT}=${component}` }}
```
Comment (nicolaserny) [inline src/components/SideBar.tsx]: you should put the pathname in a constant.
We don't want to get hard coded paths everywhere in our code

File: src/components/XCSpyMainPage.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+class XCSpyMainPage extends React.Component<XCSpyMainPageGlobalProps, XCSpyState> {
+    constructor(props: XCSpyMainPageGlobalProps) {
+        super(props);
+    }
+
+    componentWillMount() {
+        if (!this.props.initialized) {
+            const serverUrl = this.props.serverUrl;
+            const api = this.props.api;
+            sessionXCSpy.init(api, serverUrl);
```
Comment (nicolaserny) [inline src/components/XCSpyMainPage.tsx]: not sure it's the right place to init the xcomponent session. It's just a ui component

File: src/components/XCSpyMainPage.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    api: string;
+    serverUrl: string;
+    currentComponent: string;
+};
+
+interface XCSpyMainPageCallbackProps {
+    setCompositionModel: (xcApiName: string, serverUrl: string) => void;
+};
+
+class XCSpyMainPage extends React.Component<XCSpyMainPageGlobalProps, XCSpyState> {
```
Comment (nicolaserny) [inline src/components/XCSpyMainPage.tsx]: do you really need to define your component as a class ?

File: src/index.tsx [hunk 6]
```diff
@@ -20,12 +20,15 @@ import AppHeader from "components/AppHeader";
...
-import { BrowserRouter as Router, Route, NavLink } from "react-router-dom";
+import { BrowserRouter as Router, Route, Link, withRouter, Redirect } from "react-router-dom";
+import XCSpyMainPage from "components/XCSpyMainPage";
+import {API, CURRENT_COMPONENT, SERVER_URL } from "utils/urlParams";
 
 const middleware = applyMiddleware(thunk, logger());
 const store = createStore(SpyReducer, middleware);
 
 interface XCSpyGlobalProps extends XCSpyProps, XCSpyCallbackProps {
+  history: any;
```
Comment (nicolaserny) [inline src/index.tsx]: type

File: src/index.tsx [hunk 7]
```diff
@@ -92,11 +81,16 @@ const mapDispatchToProps = (dispatch: Dispatch<XCSpyState>) => {
...
 
-const App = connect(mapStateToProps, mapDispatchToProps)(XCSpyApp);
+XCSpyApp = connect(mapStateToProps, mapDispatchToProps)(XCSpyApp);
 
 ReactDOM.render(
   <Provider store={store} >
-    <App />
+    <Router>
+      <div>
+        <Route exact path="/" component={XCSpyApp} />
```
Comment (nicolaserny) [inline src/index.tsx]: Best practice: you should try to define the routes in a separate files (along with path constants).
Have a look at the command center project

File: webpack.config.js [hunk 8]
```diff
@@ -1,11 +1,13 @@
...
+const ExtractTextPlugin = require('extract-text-webpack-plugin');
 
-var BUILD_DIR = path.resolve(__dirname, "dist");
-var APP_DIR = path.resolve(__dirname, "src");
+const BUILD_DIR = path.resolve(__dirname, "dist");
+const APP_DIR = path.resolve(__dirname, "src");
 
-var config = {
+const config = {
     entry: ["es6-shim", APP_DIR + "/index.tsx"],
```
Comment (nicolaserny) [inline webpack.config.js]: replace es6-shim by babel polyfill (check command center project)
It's a fix about ie 11 / edge bugs

File: package.json (hunk 1 above)
Comment (mbarekh) [inline package.json]: done

File: src/components/Components.tsx (hunk 2 above)
Comment (mbarekh) [inline src/components/Components.tsx]: done

Comment (mbarekh) [inline src/components/Components.tsx]: compositionModel is defined in the reactiveAPI

File: src/components/SideBar.tsx (hunk 3 above)
Comment (mbarekh) [inline src/components/SideBar.tsx]: done

File: src/components/XCSpyMainPage.tsx (hunk 5 above)
Comment (mbarekh) [inline src/components/XCSpyMainPage.tsx]: ComponentWillMount is used, so we have to define it as a class

File: src/components/XCSpyMainPage.tsx (hunk 4 above)
Comment (mbarekh) [inline src/components/XCSpyMainPage.tsx]: done

File: src/index.tsx (hunk 6 above)
Comment (mbarekh) [inline src/index.tsx]: done

File: src/index.tsx (hunk 7 above)
Comment (mbarekh) [inline src/index.tsx]: done

File: webpack.config.js (hunk 8 above)
Comment (mbarekh) [inline webpack.config.js]: done

File: src/components/XCSpyMainPage.tsx [hunk 9]
```diff
@@ -1,94 +1,83 @@
...
-class XCSpyMainPage extends React.Component<XCSpyMainPageGlobalProps, XCSpyState> {
-    constructor(props: XCSpyMainPageGlobalProps) {
-        super(props);
-    }
+class XCSpyMainPage extends React.Component<XCSpyGlobalProps, XCSpyState> {
+  constructor(props: XCSpyGlobalProps) {
+    super(props);
+  }
+
+  componentWillMount() {
```
Comment (nicolaserny) [inline src/components/XCSpyMainPage.tsx]: Remove the method (nothing is done in componentWillMount)

File: src/components/XCSpyMainPage.tsx [hunk 10]
```diff
@@ -1,94 +1,83 @@
...
+interface XCSpyCallbackProps {
+  setCompositionModel: (xcApiName: string, serverUrl: string) => void;
+  initSession: (xcApiName: string, serverUrl: string) => void;
 };
 
-class XCSpyMainPage extends React.Component<XCSpyMainPageGlobalProps, XCSpyState> {
-    constructor(props: XCSpyMainPageGlobalProps) {
-        super(props);
-    }
+class XCSpyMainPage extends React.Component<XCSpyGlobalProps, XCSpyState> {
```
Comment (nicolaserny) [inline src/components/XCSpyMainPage.tsx]: Convert class in lambda function (more compact way of writing react components)

File: src/core.ts [hunk 11]
```diff
@@ -1,7 +1,17 @@
 import sessionXCSpy from "utils/sessionXCSpy";
 import { Dispatch } from "redux";
-import { updateGraphic } from "actions";
+import { updateGraphic, INIT_COMPOSITION_MODEL } from "actions";
 import { XCSpyState } from "reducers/SpyReducer";
+import xcomponentapi from "reactivexcomponent.js";
+
+export const getCompositionModel = (dispatch: Dispatch<XCSpyState>, xcApiName: string, serverUrl: string): void => {
```
Comment (nicolaserny) [inline src/core.ts]: It looks like an async action.
Should be in an action file instead of core ?

File: src/reducers/session.ts [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+export interface SessionState {
+    initialized: boolean;
+};
+
+const initialState: SessionState = {
+    initialized: false
+};
+
+export const sessionReducer: Reducer<SessionState> = (state: SessionState = initialState, action: InitSessionAction): SessionState => {
```
Comment (nicolaserny) [inline src/reducers/session.ts]: unit tests ?

File: package.json [hunk 13]
```diff
@@ -44,6 +44,7 @@
     "@types/jest": "^16.0.1",
     "@types/react": "^0.14.55",
     "@types/react-dom": "^0.14.19",
+    "@types/react-router-dom": "^4.0.4",
     "deep-freeze": "0.0.1",
     "es6-promisify": "^5.0.0",
     "es6-shim": "^0.35.2",
```
Comment (mbarekh) [inline package.json]: done

File: src/components/XCSpyMainPage.tsx (hunk 10 above)
Comment (mbarekh) [inline src/components/XCSpyMainPage.tsx]: done

File: src/components/XCSpyMainPage.tsx (hunk 9 above)
Comment (mbarekh) [inline src/components/XCSpyMainPage.tsx]: done

File: src/core.ts (hunk 11 above)
Comment (mbarekh) [inline src/core.ts]: done

Comment (mbarekh) [inline src/core.ts]: it is a xcomponent call that uses an action defined in compositionModel

File: src/reducers/session.ts (hunk 12 above)
Comment (mbarekh) [inline src/reducers/session.ts]: done

Comment (nicolaserny) [inline src/reducers/session.ts]: ?? I cannot find any SessionReducerTests in your last commit ?

Comment (mbarekh) [inline src/reducers/session.ts]: Ah i just forgot to add the file

Comment (mbarekh) [inline src/reducers/session.ts]: done

**Pull request merged**